File: public/cpp/088-task-scheduler.cpp
/*
 * Problem: Task Scheduler (LeetCode 621)
 * Link: https://leetcode.com/problems/task-scheduler/
 * Difficulty: Medium
 * Category: Heaps and Priority Queues
 *
 * Description:
 * Given a characters array tasks, representing tasks a CPU needs to do, where
 * each letter represents a different task. Tasks could be done in any order.
 * Each task is done in one unit of time. For each unit of time the CPU could
 * complete either one task or just be idle.
 * However, there is a non-negative integer n that represents the cooldown
 * interval between two same tasks — there must be at least n units of time
 * between any two same tasks.
 * Return the minimum number of intervals the CPU will take to finish all tasks.
 *
 * Example 1: tasks=["A","A","A","B","B","B"], n=2  → 8
 *   A->B->idle->A->B->idle->A->B
 * Example 2: tasks=["A","A","A","B","B","B"], n=0  → 6
 * Example 3: tasks=["A","A","A","A","A","A","B","C","D","E","F","G"], n=2 → 16
 *
 * Constraints:
 * - 1 <= tasks.length <= 10^4
 * - tasks[i] is upper-case English letter
 * - 0 <= n <= 100
 */

#include <vector>
#include <queue>
#include <unordered_map>
#include <algorithm>
#include <iostream>
using namespace std;

/*
 * APPROACH 1: MATH FORMULA (OPTIMAL, O(1) RESULT)
 *
 * maxFreq = highest task frequency
 * maxCount = number of tasks with that frequency
 * result = max(tasks.size(), (maxFreq-1)*(n+1) + maxCount)
 *
 * Intuition: the most frequent task creates "frames" of size n+1.
 * Last frame just needs maxCount slots. Fill rest with other tasks or idle.
 *
 * Time:  O(|tasks|)
 * Space: O(1)  (26-char frequency array)
 */
class Solution1 {
public:
    int leastInterval(vector<char>& tasks, int n) {
        int freq[26] = {};
        for (char c : tasks) freq[c - 'A']++;
        int maxFreq = *max_element(freq, freq + 26);
        int maxCount = count(freq, freq + 26, maxFreq);
        return max((int)tasks.size(), (maxFreq - 1) * (n + 1) + maxCount);
    }
};

/*
 * APPROACH 2: MAX HEAP + COOLDOWN QUEUE SIMULATION
 *
 * Use a max-heap of (freq, task). Each cycle of n+1 slots:
 * - Pop up to n+1 tasks from heap, decrement freq
 * - Push back tasks with remaining freq > 0
 * - Add idle time if fewer than n+1 tasks were available
 *
 * Time:  O(|tasks| * n)
 * Space: O(1)  (26 unique tasks at most)
 */
class Solution2 {
public:
    int leastInterval(vector<char>& tasks, int n) {
        int freq[26] = {};
        for (char c : tasks) freq[c - 'A']++;
        priority_queue<int> pq;
        for (int f : freq) if (f) pq.push(f);

        int time = 0;
        while (!pq.empty()) {
            vector<int> tmp;
            int cycle = n + 1;
            while (cycle-- && !pq.empty()) {
                int f = pq.top(); pq.pop();
                if (f > 1) tmp.push_back(f - 1);
                time++;
            }
            for (int f : tmp) pq.push(f);
            if (!pq.empty()) time += cycle + 1; // idle slots
        }
        return time;
    }
};

/*
 * APPROACH 3: GREEDY SIMULATION WITH QUEUE
 *
 * Simulate time step-by-step with a max-heap and a cooldown queue of
 * (freq, available_at) pairs. Closest to real scheduler behaviour.
 *
 * Time:  O(|tasks| log 26) = O(|tasks|)
 * Space: O(1)
 */
class Solution3 {
public:
    int leastInterval(vector<char>& tasks, int n) {
        int freq[26] = {};
        for (char c : tasks) freq[c - 'A']++;
        priority_queue<int> pq;
        for (int f : freq) if (f) pq.push(f);

        // cooldown queue: (remaining_count, next_available_time)
        queue<pair<int,int>> cooldown;
        int time = 0;

        while (!pq.empty() || !cooldown.empty()) {
            time++;
            // release tasks whose cooldown has expired
            if (!cooldown.empty() && cooldown.front().second <= time) {
                pq.push(cooldown.front().first);
                cooldown.pop();
            }
            if (!pq.empty()) {
                int f = pq.top(); pq.pop();
                if (f - 1 > 0) cooldown.push({f - 1, time + n + 1});
            }
        }
        return time;
    }
};

/*
 * APPROACH 4: COUNTING IDLE SLOTS
 *
 * Explicitly count idle slots needed:
 * idle = (maxFreq-1)*n - sum of min(maxFreq-1, freq) for other tasks
 * result = tasks.size() + max(0, idle)
 *
 * Time:  O(|tasks|)
 * Space: O(1)
 */
class Solution4 {
public:
    int leastInterval(vector<char>& tasks, int n) {
        int freq[26] = {};
        for (char c : tasks) freq[c - 'A']++;
        sort(freq, freq + 26);
        int maxFreq = freq[25];
        int idle = (maxFreq - 1) * n;
        for (int i = 24; i >= 0 && idle > 0; i--)
            idle -= min(maxFreq - 1, freq[i]);
        idle = max(0, idle);
        return (int)tasks.size() + idle;
    }
};

int main() {
    auto run = [](vector<char> tasks, int n) {
        cout << "tasks.size=" << tasks.size() << " n=" << n << "\n";
        { Solution1 s; cout << "Formula:    " << s.leastInterval(tasks,n) << "\n"; }
        { Solution2 s; cout << "HeapCycle:  " << s.leastInterval(tasks,n) << "\n"; }
        { Solution3 s; cout << "GreedySim:  " << s.leastInterval(tasks,n) << "\n"; }
        { Solution4 s; cout << "IdleCount:  " << s.leastInterval(tasks,n) << "\n"; }
        cout << "\n";
    };
    run({'A','A','A','B','B','B'}, 2);   // → 8
    run({'A','A','A','B','B','B'}, 0);   // → 6
    run({'A','A','A','A','A','A','B','C','D','E','F','G'}, 2); // → 16
    return 0;
}

/*
 * COMPLEXITY SUMMARY:
 * Approach 1 (Formula):      Time O(N)   Space O(1)  ← optimal
 * Approach 2 (HeapCycle):    Time O(N*n) Space O(1)
 * Approach 3 (GreedySim):    Time O(N)   Space O(1)
 * Approach 4 (IdleCount):    Time O(N)   Space O(1)
 */

// Made with Bob

File: src/data/tutorials/088-task-scheduler.js
/**
 * Tutorial content for #088 — Task Scheduler
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `Given a list of tasks (each represented by a letter) and a non-negative cooldown integer n, find the minimum number of CPU intervals needed to finish all tasks. Between any two identical tasks there must be at least n intervals (tasks or idle slots).`,
    example: `tasks = ["A","A","A","B","B","B"], n = 2\nSchedule: A B _ A B _ A B\nEach A is at least 2 slots apart, each B too.\nTotal intervals = 8\n✅ Answer: 8`,
    keyInsight: `The most frequent task determines the minimum number of "frames" needed. Each frame is (n+1) slots. Extra tasks fill the idle slots. If all tasks fill slots completely, the answer is simply len(tasks).`,
  },

  approaches: {
    'Math Formula': {
      intuition: `Let maxFreq = the frequency of the most common task. Let maxCount = how many tasks share that maximum frequency.\n\nThe minimum time is determined by the "frame" structure around the most frequent task:\n  frames = maxFreq - 1\n  slots per frame = n + 1\n  extra slots at the end = maxCount (tasks with same max freq that go in the last frame)\n\nSo: answer = max(tasks.length, (maxFreq - 1) * (n + 1) + maxCount)`,
      steps: [
        `Count the frequency of each task.`,
        `Find maxFreq = maximum frequency.`,
        `Find maxCount = number of tasks with frequency == maxFreq.`,
        `Return max(tasks.length, (maxFreq - 1) * (n + 1) + maxCount).`,
      ],
      example: `tasks=["A","A","A","B","B","B"], n=2\n\nfreqs: A→3, B→3. maxFreq=3, maxCount=2.\n(maxFreq-1)*(n+1)+maxCount = (3-1)*(2+1)+2 = 2*3+2 = 8\nmax(6, 8) = 8 ✅\n\ntasks=["A","A","A","B","B","B","C","C","C","D","D","E"], n=2\nmaxFreq=3, maxCount=3 (A,B,C)\n(3-1)*3+3=9. tasks.length=12. max(12,9)=12 ✅`,
      keyInsight: `O(n) time, O(1) space (only 26 possible tasks). This single formula is the key insight of the problem — no simulation needed. The hardest part is understanding why the formula works.`,
    },

    'Max Heap Cycle Simulation': {
      intuition: `Use a max-heap keyed by frequency. In each "cycle" of length (n+1), greedily pick up to (n+1) tasks from the heap (always the most frequent remaining task). After the cycle, push back tasks that still have remaining count. Count idle slots when fewer than (n+1) tasks are available.`,
      steps: [
        `Count frequencies, build max-heap.`,
        `time = 0.`,
        `While heap not empty:`,
        `  temp = [] (tasks used this cycle).`,
        `  For i = 0 to n (up to n+1 slots):`,
        `    If heap not empty: pop (freq, task); temp.add((freq-1, task)).`,
        `  Push back entries from temp with freq > 0.`,
        `  time += heap is empty after this cycle ? temp size (tasks done) : n + 1.`,
        `Return time.`,
      ],
      example: `tasks=["A","A","A","B","B","B"], n=2\nheap: [(3,A),(3,B)]\n\nCycle 1: pop A(3→2), pop B(3→2), 1 idle slot. time+=3. heap:[(2,A),(2,B)]\nCycle 2: pop A(2→1), pop B(2→1), idle. time+=3. heap:[(1,A),(1,B)]\nCycle 3: pop A(1→0), pop B(1→0). heap empty → time+=2 (no idle).\ntotal = 3+3+2 = 8 ✅`,
      keyInsight: `O(n log 26) = O(n) time (heap has at most 26 entries). Good for visualizing the actual schedule. The Math Formula is faster to implement; this is better for understanding why the schedule works.`,
    },

    'Greedy Simulation with Cooldown Queue': {
      intuition: `Simulate the CPU timeline explicitly. Maintain a "cooldown queue" of tasks that were recently used and can't be used yet (with the timestamp when they become available). At each tick, pick the most frequent available task, or idle if none is available.`,
      steps: [
        `Count frequencies, build max-heap.`,
        `Queue of (freq, availableTime) for tasks on cooldown.`,
        `time = 0.`,
        `While heap or queue is not empty:`,
        `  time++.`,
        `  If heap not empty: pop most frequent; if freq-1>0, add to queue with availableTime=time+n.`,
        `  If queue.front.availableTime == time: move back to heap.`,
        `Return time.`,
      ],
      example: `tasks=["A","A","A","B","B","B"], n=2\n\nt=1: pop A(3→2). queue:[(2,A,t=3)]\nt=2: pop B(3→2). queue:[(2,A,3),(2,B,4)]\nt=3: heap empty. idle. release A → heap:[(2,A),(0,B? wait)]\nt=3: A available → pop A(2→1). queue:[(2,B,4),(1,A,5)]\n... (continues)\nTotal time = 8 ✅`,
      keyInsight: `O(n log 26) = O(n) time. This approach explicitly models the scheduler and is very intuitive — you can trace the exact schedule. Slightly more complex to code than the Math Formula.`,
    },

    'Counting Idle Slots': {
      intuition: `Calculate the number of idle slots directly. Start with (maxFreq - 1) * n idle slots (one frame of n slots between each occurrence of the most frequent task). Then fill those idle slots by inserting other tasks. If idle slots go negative, there are no idles needed.`,
      steps: [
        `Count task frequencies.`,
        `Sort frequencies descending.`,
        `maxFreq = freqs[0].`,
        `idleSlots = (maxFreq - 1) * n.`,
        `For each remaining freq (freqs[1..]):`,
        `  idleSlots -= min(freq, maxFreq - 1).`,
        `Return tasks.length + max(0, idleSlots).`,
      ],
      example: `tasks=["A","A","A","B","B","B"], n=2\nfreqs=[3,3]. maxFreq=3. idleSlots=(3-1)*2=4.\n\nFor B (freq=3): idleSlots -= min(3, 3-1=2) = 4-2 = 2.\nRemaining idleSlots = 2.\nAnswer = 6 + max(0,2) = 8 ✅`,
      keyInsight: `O(n + 26 log 26) = O(n) time. A different angle on the same math formula — directly counting idle slots and filling them with other tasks. Equivalent result to the Math Formula.`,
    },
  },
}
